### PR TITLE
Kerberos smart card credentials handling

### PR DESCRIPTION
**Pull request by TheBestTvarynka**
Hello,
In this PR I've added Kerberos smart card credentials handling. _Note_: it doesn't contain actual smart card logon implementation or code for working with smart cards (they will be in other PRs).

Changes overview:

* Added the "_scard_" feature to enable the smart card logon. This feature can be enabled only on Windows.
* Improved the `ffi` bindings. Now users can pass the smart card credentials using the `AcquireCredentialsHandleA/W` function. How it works: the smart card credentials are always marshaled into a string that always starts with the "@" character.
* Added a `CredentialsBuffers`, `Credentials`, `SmartCardIdentityBuffers`, and `SmartCardIdentity` structure to be able to handle any kind of credentials.
* I didn't change the `credentials_handle` type of the CredSSP client because of backward compatibility with the IronRDP that doesn't support smart card logon yet.
* Improved TsCredentials encoding. For the `TsCredentials`, `TsSmartCardCreds`, and `TsCspDataDetail` structures I use the `picky-krb` crate. See the corresponding PR: https://github.com/Devolutions/picky-rs/pull/233.

Docs & references:

* [MS-CSSP: TSCredentials](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-cssp/94a1ab00-5500-42fd-8d3d-7a84e6c2cf03).
* [CredUnmarshalCredentialW](https://learn.microsoft.com/en-us/windows/win32/api/wincred/nf-wincred-credunmarshalcredentialw).

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -46,11 +48,13 @@ serde = "1.0"
...
-picky-krb = "0.6.0"
-picky-asn1 = { version = "0.7.1", features = ["chrono_conversion"] }
-picky-asn1-der = "0.4.0"
-picky-asn1-x509 = { version = "0.9.0", features = ["pkcs7"] }
+
+picky = { git = "https://github.com/Devolutions/picky-rs.git", rev = "fda2fad12a704f2d5008f1ec4df62e5ca8a5e8ff", default-features = false }
+picky-asn1 = { git = "https://github.com/Devolutions/picky-rs.git", rev = "fda2fad12a704f2d5008f1ec4df62e5ca8a5e8ff", features = ["chrono_conversion"] }
+picky-asn1-der = { git = "https://github.com/Devolutions/picky-rs.git", rev = "fda2fad12a704f2d5008f1ec4df62e5ca8a5e8ff" }
+picky-asn1-x509 = { git = "https://github.com/Devolutions/picky-rs.git", rev = "fda2fad12a704f2d5008f1ec4df62e5ca8a5e8ff", features = ["pkcs7"] }
+picky-krb = { git = "https://github.com/Devolutions/picky-rs.git", rev = "fda2fad12a704f2d5008f1ec4df62e5ca8a5e8ff" }
```
Comment (TheBestTvarynka) [inline Cargo.toml:56]: I will update the `picky-*` dependencies once https://github.com/Devolutions/picky-rs/pull/233 is merged and the new versions of the crates are published

Comment (CBenoit) [reply Cargo.toml:56]: Sounds good. Newer picky versions will be published in the next few days, at least before Friday :+1: (waiting for a pull request to get merged)

File: ffi/src/sec_winnt_auth_identity.rs [hunk 2]
```diff
@@ -158,25 +162,11 @@ pub unsafe fn get_auth_data_identity_version_and_flags(p_auth_data: *const c_voi
...
-    package_list: &mut Option<String>,
-) -> Result<AuthIdentityBuffers> {
-    let (_, auth_flags) = get_auth_data_identity_version_and_flags(p_auth_data);
-
-    if (auth_flags & SEC_WINNT_AUTH_IDENTITY_UNICODE) != 0 {
-        auth_data_to_identity_buffers_w(security_package_name, p_auth_data, package_list)
-    } else {
-        auth_data_to_identity_buffers_a(security_package_name, p_auth_data, package_list)
-    }
-}
```
[the remark this answers is not in the document]
Comment (CBenoit) [reply ffi/src/sec_winnt_auth_identity.rs:173]: Sounds good.

File: ffi/Cargo.toml [hunk 3]
```diff
@@ -15,6 +15,7 @@ crate-type = ["cdylib"]
 [features]
 default = []
 tsssp = ["sspi/tsssp"]
+scard = ["sspi/scard"]
```
Comment (CBenoit) [inline ffi/Cargo.toml:18]: I’m not sure if this should be feature-gated for the DLL. Don’t we want to package the smart card login support for all Windows builds?
Any input @awakecoding

**Review by CBenoit — COMMENTED** [review]
Looking good.

> I didn't change the `credentials_handle` type of the CredSSP client because of backward compatibility with the IronRDP that doesn't support smart card logon yet.

You can introduce a breaking change if necessary. We’ll increase the version number as appropriate and fix downstream code (IronRDP). I guess it would not be too hard in this case.

> Improved the ffi bindings. Now users can pass the smart card credentials using the AcquireCredentialsHandleA/W function. How it works: the smart card credentials are always marshaled into a string that always starts with the "@" character.

Any opinion? @awakecoding

File: ffi/src/sec_winnt_auth_identity.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheBestTvarynka) [reply ffi/src/sec_winnt_auth_identity.rs:173]: Where should I add the comment? I mean the function is deleted, so...

File: Cargo.toml (hunk 1 above)
Comment (TheBestTvarynka) [reply Cargo.toml:56]: I have two more pull requests in the scope of the Kerberos smart card logon. Maybe we can merge this PR as it is? Anyway, we merge it in the dev branch and I can fix it later. I just want to minimize the code review time

**Comment by TheBestTvarynka** [comment]
I improved the `CredSspClient/Server`. Now it uses a `Credentials` structure and can handle both types of credentials.

File: ffi/src/sec_winnt_auth_identity.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (CBenoit) [reply ffi/src/sec_winnt_auth_identity.rs:173]: Above `auth_data_to_identity_buffers_w` and `auth_data_to_identity_buffers_a` sounds good to me

File: Cargo.toml (hunk 1 above)
Comment (CBenoit) [reply Cargo.toml]: It won’t be necessary, new picky versions are incoming in the next few hours

Comment (CBenoit) [reply Cargo.toml]: Sorry for the delay. The new versions are available!

File: ffi/Cargo.toml (hunk 3 above)
Comment (CBenoit) [reply ffi/Cargo.toml:18]: After discussing with Marc-André, we agreed to keep this behind a feature gate. The main motivation is to not include the new smart card code into production build while it’s not ready yet.

File: Cargo.toml (hunk 1 above)
Comment (TheBestTvarynka) [reply Cargo.toml]: I updated `picky-*` dependencies

File: ffi/src/sec_winnt_auth_identity.rs [hunk 4]
```diff
@@ -263,7 +265,7 @@ pub unsafe fn auth_data_to_identity_buffers_w(
 
         // only marshaled smart card creds starts with '@' char
         #[cfg(feature = "scard")]
-        if user[0] == b'@' {
+        if user[0] == b'@' && CredIsMarshaledCredentialW(user.as_ptr() as *const _) != 0 {
```
Comment (awakecoding) [inline ffi/src/sec_winnt_auth_identity.rs:268]: calling CredIsMarshaledCredentialW should be enough, it checks for @@ under the hood anyway. maybe the only potential issue is that CredIsMarshaledCredentialW could return true for types other than 1

Comment (TheBestTvarynka) [reply ffi/src/sec_winnt_auth_identity.rs]: done

**Comment by CBenoit** [comment]
This is looking good to me. Do we need to publish a new version of the Rust crate, or should this wait for the follow-up PRs?

**Comment by TheBestTvarynka** [comment]
Wait for follow up prs. I have 2 more pull requests

**Review by CBenoit — APPROVED** [review]

**Pull request merged**